ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala

```diff
@@ -128,104 +138,120 @@ private[events] class BufferedTransactionsReader(
+      executionContext: ExecutionContext
   ): BufferedTransactionsReader =
     new BufferedTransactionsReader(
       delegate = delegate,
       transactionsBuffer = transactionsBuffer,
-      toFlatTransaction =
-        TransactionLogUpdatesConversions.ToFlatTransaction(_, _, _, _, _, lfValueTranslation),
-      toTransactionTree =
-        TransactionLogUpdatesConversions.ToTransactionTree(_, _, _, lfValueTranslation),
       metrics = metrics,
+      filterFlatTransactions = ToFlatTransaction.filter,
+      flatToApiTransactions =
+        ToFlatTransaction.toApiTransaction(_, _, lfValueTranslation)(_, executionContext),
+      filterTransactionTrees = ToTransactionTree.filter,
+      treesToApiTransactions =
+        ToTransactionTree.toApiTransaction(_, _, lfValueTranslation)(_, executionContext),
     )
 
-  private[events] def getTransactions[FILTER, API_TX, API_RESPONSE](
-      transactionsBuffer: EventsBuffer[Offset, TransactionLogUpdate]
+  private[events] def getTransactions[FILTER, API_RESPONSE](
+      transactionsBuffer: EventsBuffer[TransactionLogUpdate]
   )(
       startExclusive: Offset,
       endInclusive: Offset,
       filter: FILTER,
       verbose: Boolean,
+      metrics: Metrics,
   )(
-      toApiTx: (TxUpdate, FILTER, Boolean) => Future[Option[API_TX]],
-      apiResponseCtor: Seq[API_TX] => API_RESPONSE,
+      filterEvents: Filter,
+      toApiTx: ToApi[API_RESPONSE],
       fetchTransactions: FetchTransactions[FILTER, API_RESPONSE],
-      sourceTimer: Timer,
-      toApiTxTimer: Timer,
-      resolvedFromBufferCounter: Counter,
-      totalRetrievedCounter: Counter,
-      outputStreamBufferSize: Int,
-      bufferSizeCounter: Counter,
-  )(implicit executionContext: ExecutionContext): Source[(Offset, API_RESPONSE), NotUsed] = {
-    def bufferedSource(
-        slice: Vector[(Offset, TransactionLogUpdate)]
-    ): Source[(Offset, API_RESPONSE), NotUsed] =
-      Source
-        .fromIterator(() => slice.iterator)
-        // Using collect + mapAsync as an alternative to the non-existent collectAsync
-        .collect { case (offset, tx: TxUpdate) =>
-          Timed.future(toApiTxTimer, toApiTx(tx, filter, verbose).map(offset -> _))
-        }
-        // Note that it is safe to use high parallelism for mapAsync as long
-        // as the Futures executed within are running on a bounded thread pool
-        .mapAsync(32)(identity)
-        .async
-        .collect { case (offset, Some(tx)) =>
-          resolvedFromBufferCounter.inc()
-          offset -> apiResponseCtor(Seq(tx))
-        }
+      bufferReaderMetrics: metrics.daml.services.index.BufferedReader,
+  ): Source[(Offset, API_RESPONSE), NotUsed] = {
+    def bufferSource(
+        bufferSlice: Vector[(Offset, TransactionLogUpdate.Transaction)]
+    ) =
+      if (bufferSlice.isEmpty) Source.empty
+      else
+        Source
+          .fromIterator(() => bufferSlice.iterator)
+          .mapAsync(1) { case (offset, payload) =>
```

Comment:
A general remark what we discussed offline: I am slightly concerned whether we will know if this transformation is the bottleneck for stream delivery. As implementing the ACS 2 phase retrieval, I was making tests without verbose mode, an d with parallelism 2 it was never a bottleneck, but this might look very differently here.

In the ACS implementation in TransactionReader this is coming from a static config (set to two). I am wondering whether we should do here the same.

(of course this only helps in a logical way if the toApiTx is async, see my corresponding comments in the conversions file, next to the implementation. There I also elaborate why it cannot always be expected the implementation to run in parallel at all for one such call)